core/preconditioner/isai_kernels.hpp

```diff
@@ -42,27 +42,37 @@ namespace gko {
-    void generate_l_inverse(std::shared_ptr<const DefaultExecutor> exec,    \
-                            const matrix::Csr<ValueType, IndexType> *l_csr, \
-                            matrix::Csr<ValueType, IndexType> *inverse_l)
-
-#define GKO_DECLARE_ISAI_GENERATE_U_INVERSE_KERNEL(ValueType, IndexType)    \
-    void generate_u_inverse(std::shared_ptr<const DefaultExecutor> exec,    \
-                            const matrix::Csr<ValueType, IndexType> *u_csr, \
-                            matrix::Csr<ValueType, IndexType> *inverse_u)
-
-#define GKO_DECLARE_ISAI_IDENTITY_TRIANGLE_KERNEL(ValueType, IndexType) \
-    void identity_triangle(std::shared_ptr<const DefaultExecutor> exec, \
-                           matrix::Csr<ValueType, IndexType> *mtx, bool lower)
-
-#define GKO_DECLARE_ALL_AS_TEMPLATES                                  \
-    template <typename ValueType, typename IndexType>                 \
-    GKO_DECLARE_ISAI_GENERATE_L_INVERSE_KERNEL(ValueType, IndexType); \
-    template <typename ValueType, typename IndexType>                 \
-    GKO_DECLARE_ISAI_GENERATE_U_INVERSE_KERNEL(ValueType, IndexType); \
-    template <typename ValueType, typename IndexType>                 \
-    GKO_DECLARE_ISAI_IDENTITY_TRIANGLE_KERNEL(ValueType, IndexType)
+#define GKO_DECLARE_ISAI_GENERATE_TRI_INVERSE_KERNEL(ValueType, IndexType)    \
+    void generate_tri_inverse(std::shared_ptr<const DefaultExecutor> exec,    \
+                              const matrix::Csr<ValueType, IndexType> *input, \
+                              matrix::Csr<ValueType, IndexType> *inverse,     \
+                              IndexType *excess_rhs_ptrs,                     \
+                              IndexType *excess_nz_ptrs, bool lower)
+
+#define GKO_DECLARE_ISAI_GENERATE_EXCESS_SYSTEM_KERNEL(ValueType, IndexType) \
+    void generate_excess_system(                                             \
+        std::shared_ptr<const DefaultExecutor> exec,                         \
+        const matrix::Csr<ValueType, IndexType> *input,                      \
+        const matrix::Csr<ValueType, IndexType> *inverse,                    \
+        const IndexType *excess_rhs_ptrs, const IndexType *excess_nz_ptrs,   \
+        matrix::Csr<ValueType, IndexType> *excess_system,                    \
+        matrix::Dense<ValueType> *excess_rhs)
+
+#define GKO_DECLARE_ISAI_SCATTER_EXCESS_SOLUTION_KERNEL(ValueType, IndexType) \
+    void scatter_excess_solution(                                             \
+        std::shared_ptr<const DefaultExecutor> exec,                          \
+        const IndexType *excess_rhs_ptrs,                                     \
+        const matrix::Dense<ValueType> *excess_solution,                      \
+        matrix::Csr<ValueType, IndexType> *inverse)
+
+#define GKO_DECLARE_ALL_AS_TEMPLATES                                      \
+    constexpr auto row_size_limit = 32;                                   \
```

Comment:
I dislike the placement of this variable because I looked for its definition, and it took me some time to find it here.
I would prefer if it was placed in the `cuda`, `hip`, ... kernel files since all other variables are also defined there. This might lead to a bit of code duplication, but makes it easier to fine-tune them for each executor and also easier to find.

Reply:
The reason I need the same value for each executor is that otherwise it is impossible to compare the individual kernel results with Reference. The only other option I see would be making the row size limit a parameter for the kernel, but that would go against the limits imposed by the hardware.

Reply:
That does make sense, I currently can't come up with a better solution, so I will mark this as resolved. It is still not ideal IMO.